RHDEVDOCS-3226 Logging 5.2 docs aggregated branch (#35445)
openshift/openshift-docs: modules/cluster-logging-collector-log-forward-loki.adoc

= Forwarding logs to Grafana Loki

You can forward logs to an external Loki logging system in addition to, or instead of, the internal default {product-title} Elasticsearch instance.

To configure log forwarding to Grafana Loki, create a `ClusterLogForwarder` custom resource (CR) with an output to Loki and a pipeline that uses the output. The output to Loki can use the HTTP (insecure) or HTTPS (secure HTTP) connection.

.Prerequisites

* You must have a Loki logging system running at the URL you specify with the `url` field in the custom resource.

.Procedure

. Create a `ClusterLogForwarder` CR YAML file similar to the following:
+
[source,yaml]
----
apiVersion: "logging.openshift.io/v1"
kind: ClusterLogForwarder
metadata:
  name: instance <1>
  namespace: openshift-logging <2>
spec:
  outputs:
   - name: loki-insecure <3>
     type: "loki" <4>
     url: http://loki.insecure.com:9200 <5>
   - name: loki-secure
     type: "loki"
     url: https://loki.secure.com:9200 <6>
     secret:
        name: loki-secret <7>
  pipelines:
   - name: application-logs <8>
     inputRefs: <9>
     - application
     - audit
     outputRefs:
     - loki-secure <10>

>      tenantKey: kubernetes.namespace_name <11>
>      labelKeys: kubernetes.labels.foo   <12>

Reviewer: ```suggestion
     loki:
         tenantKey: kubernetes.namespace_name <11>
         labelKeys: kubernetes.labels.foo   <12>
```
The loki specific entries go under a `loki` field

> ----

Author: @alanconway Please review lines 44-45, above.

Reviewer: See diff.

<7> For an `https` prefix, specify the name of the secret required by the endpoint for TLS communication. The secret must exist in the `openshift-logging` project, and must have keys of: *tls.crt*, *tls.key*, and *ca-bundle.crt* that point to the respective certificates that they represent. Otherwise, for `http` and `https` prefixes, you can specify a secret that contains a username and password. For more information, see the following "Example: Setting secret that contains a username and password."
<8> Optional: Specify a name for the pipeline.
<9> Specify which log types to forward by using the pipeline: `application,` `infrastructure`, or `audit`.
<10> Specify the name of the output to use when forwarding logs with this pipeline.
<11> Optional: Specify a meta-data key field to generate values for the `TenantID` field in Loki. For example, setting `tenantKey: kubernetes.namespace_name` uses the names of the Kubernetes namespaces as values for tenant IDs in Loki. To see which other log record fields you can specify, see the "Log Record Fields" link in the following "Additional Resources" section.

> <12> Optional: Specify a list of meta-data field keys to replace the default Loki labels. Loki label names must match the regular expression `[a-zA-Z_:][a-zA-Z0-9_:]*`. Illegal characters in meta-data keys are replaced with `_` to form the label name. For example the `kubernetes.labels.foo` meta-data key becomes Loki label `kubernetes_labels_foo`. If you do not set `labelKeys`, the default labels are: `[log_type, kubernetes_namespace_name, kubernetes_pod_name, kubernetes_host]`. Keep the set of labels small because Loki limits the size and number of labels allowed. See link:https://grafana.com/docs/loki/latest/configuration/#limits_config[Configuring Loki, limits_config]. You can still query based on any log record field using query filters.

Reviewer: ```suggestion
<12> Optional: Specify a list of meta-data field keys to replace the default Loki labels. Loki label names must match the regular expression `[a-zA-Z_:][a-zA-Z0-9_:]*`. Illegal characters in meta-data keys are replaced with `_` to form the label name. For example the `kubernetes.labels.foo` meta-data key becomes Loki label `kubernetes_labels_foo`. If you do not set `labelKeys`, the default value is: `[log_type, kubernetes.namespace_name, kubernetes.pod_name, kubernetes_host]`. Keep the set of labels small because Loki limits the size and number of labels allowed. See link:https://grafana.com/docs/loki/latest/configuration/#limits_config[Configuring Loki, limits_config]. You can still query based on any log record field using query filters.
```

The labelKeys value (like other YAML field ending in `Key` or `Keys`) uses dotted field notation, best to stick with that in this discussion (probably this is cut/paste from something I wrote, my bad) The loki label underscore notation is never used in the YAML file, it will be used when the user wants to write LogQL queries to retrieve logs from Loki.

+
[NOTE]
====
Because Loki requires log streams to be correctly ordered by timestamp, `labelKeys` always includes the `kubernetes_host` label set, even you do not specify it. This inclusion ensures that each stream originates from a single host, which prevents timestamps from becoming disordered due to clock differences on different hosts.
====

> 

Author: @alanconway Please review lines 57-62, above. Thanks.

> // Module included in the following assemblies:

Author: For Loki support - see this section here.

Author: Closed https://github.com/openshift/openshift-docs/pull/35630

Author: @alanconway

Author: @abrennan89 Please review this file.

//
// * logging/cluster-logging-external.adoc

[id="cluster-logging-collector-log-forward-loki_{context}"]
= Forwarding logs to Grafana Loki

You can forward logs to an external Loki logging system in addition to, or instead of, the internal default {product-title} Elasticsearch instance.

To configure log forwarding to Grafana Loki, create a `ClusterLogForwarder` custom resource (CR) with an output to Loki and a pipeline that uses the output. The output to Loki can use the HTTP (insecure) or HTTPS (secure HTTP) connection.

.Prerequisites

> * You must have an Loki logging system that is configured to receive the logging data using the specified protocol or format.

Reviewer: ```suggestion
* You must have a Loki logging system running at the URL specified in the custom resource.
```
There isn't anything to configure on the Loki side regarding protocol or format, we use Loki's REST API to forward logs.

Author: Updated. Thanks.

.Procedure

. Create a `ClusterLogForwarder` CR YAML file similar to the following:
+
[source,yaml]
----
apiVersion: "logging.openshift.io/v1"
kind: ClusterLogForwarder
metadata:
  name: instance <1>
  namespace: openshift-logging <2>
spec:
  outputs:
   - name: loki-insecure <3>
     type: "loki" <4>
     url: http://loki.insecure.com:9200 <5>
   - name: loki-secure
     type: "loki"
     url: https://loki.secure.com:9200 <6>
     secret:
        name: loki-secret <7>
  pipelines:
   - name: application-logs <8>
     inputRefs: <9>
     - application
     - audit
     outputRefs:
     - loki-secure <10>

>      labels:

Reviewer: I don't think we should overload the loki example with features unrelated to loki - pipeline labels can be used with any output type and are hopefully documented elsewhere?

Author: Updated. I removed everything below `- loki-secure <10>`. PTAL to make sure this looks good to you @alanconway

<8> Optional: Specify a name for the pipeline.
<9> Specify which log types to forward by using the pipeline: `application,` `infrastructure`, or `audit`.
<10> Specify the name of the output to use when forwarding logs with this pipeline.
<11> Optional: String. One or more labels to add to the logs.
<12> Optional: Configure multiple outputs to forward logs to other external log aggregators of any supported type:
** A name to describe the pipeline.
** The `inputRefs` is the log type to forward by using the pipeline: `application,` `infrastructure`, or `audit`.
** The `outputRefs` is the name of the output to use.
** Optional: String. One or more labels to add to the logs.

. Create the CR object:
+
[source,terminal]

> ----

Reviewer: There are a couple of additional loki-specific properties documented here: 
    https://github.com/openshift/cluster-logging-operator/blob/master/pkg/apis/logging/v1/output_types.go#L181
Ideally we should be generating reference doc directly from `_types.go` but for now it's a cut/paste job.

Author: @alanconway  Please review the updated contents.

$ oc create -f <file-name>.yaml
----

> The Red Hat OpenShift Logging Operator redeploys the Fluentd pods. If the pods do not redeploy, you can delete the Fluentd pods to force them to redeploy.

Reviewer: Again this is a generic troubleshooting tip not related to Loki, not sure it belongs here since: 
a) it shouldn't happen, we're suggesting the product is buggy and
b) it applies any time you update the forwarder CR for any reason, it's not specific to using Loki, so should be documented in some more general place.

Author: I agree with this proposal and am getting feedback from the team. This removal will affect about eight files.

Author: Relocated that statement to a troubleshooting section. Thanks.
